navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,163 @@
+package com.mapbox.services.android.navigation.v5.instruction;
+
+import android.text.TextUtils;
+import android.util.SparseArray;
+
+import com.mapbox.services.android.navigation.v5.NavigationConstants;
+import com.mapbox.services.android.navigation.v5.RouteProgress;
+import com.mapbox.services.api.directions.v5.models.StepManeuver;
+import com.mapbox.services.api.utils.turf.TurfConstants;
+import com.mapbox.services.api.utils.turf.TurfHelpers;
+
+import java.text.DecimalFormat;
+import java.util.Locale;
+
+class DefaultInstructionEngine extends SparseArray<DefaultInstructionEngine.InstructionBuilder> {
+
+  private static final double MINIMUM_UPCOMING_STEP_DISTANCE = 15d;
+  private static final String DECIMAL_FORMAT = "###.#";
+  private static final String MILES_STRING_FORMAT = "%s Miles";
+  private static final String FEET_STRING_FORMAT = "%s Feet";
+  private static final String IN_STRING_FORMAT = "In %s %s";
+  private static final String THEN_STRING_FORMAT = "%s then %s";
+  private static final String THEN_IN_STRING_FORMAT = "%s then in %s %s";
+  private static final String CONTINUE_STRING_FORMAT = "Continue on %s for %s";
+
+  DefaultInstructionEngine() {
+    initDefaultBuilders();
+  }
+
+  private void initDefaultBuilders() {
+    this.put(NavigationConstants.DEPARTURE_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildDepartureInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.NEW_STEP_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildNewStepInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.IMMINENT_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildImminentInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.URGENT_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildUrgentInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.ARRIVAL_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        return buildArrivalInstruction(routeProgress);
+      }
+    });
+  }
+
+  private String buildDepartureInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep().getDistance() > MINIMUM_UPCOMING_STEP_DISTANCE) {
```

Comment:
@cammace can you check my logic here?  I had interpreted it as

 _if the next step > 15 meters, announce normally, if < 15 meters, use the combo announcement_